torch/distributed/tensor/parallel/input_reshard.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import contextlib
+from functools import partial
+from typing import Any, Iterator, Optional, Tuple
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor, Replicate, Shard
+
+__all__ = [
+    "input_reshard_wrapper",
+]
+
+
+class InputReshard(torch.nn.Module):
+    """
+    An nn.Module that wraps another nn.Module with input resharding.
```

Comment:
Can't this be done with fwd hooks instead of a wrapper?

Reply:
But within it, it's a context manager, is it possible we can put that inside a fwd hook? cc: @albanD

Reply:
NVM I made it as two hooks.